### DISCUSSION TITLE
how to load  custom model for inference

### DISCUSSION THREAD

**Discussion by e101sg** [ultralytics/yolov5#5872] [Q&A] [answered]
I am new to PyTorch and training  for custom object detection. After model created , trying to load from local folder. 
```
#model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)
        
model = torch.hub.load('/Users/macmini04/Documents/yolov5','custom', path='best.pt',force_reload=True,source='local', pretrained =Flase)
```
The above code does not create bounding box on the  output video file. I am following this code  below and modify it, to load local model.  Not sure, what i am missing. Any thoughts Welcome 👍 .
[https://github.com/akash-agni/Real-Time-Object-Detection/blob/main/Drone_Human_Detection_Model.py](url)

**Comment by glenn-jocher** [comment, marked as answer]
@e101sg you have a typo in your arguments, Flase.

In any case, I don't know why you would attempt to pass pretrained=False if you are expecting trained results.

**Reply by e101sg** [reply]
oh! it is mistake.  i have corrected later but still not detection.

**Reply by glenn-jocher** [reply]
@e101sg just follow the **YOLOv5 PyTorch Hub** tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://github.com/ultralytics/yolov5/wiki/Tips-for-Best-Training-Results)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://github.com/ultralytics/yolov5/issues/2518)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://github.com/ultralytics/yolov5/issues/251) 🚀
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Reply by chef-SAbhinay** [reply]
> oh! it is mistake.  i have corrected later but still not detection. 

is this resolved?

**Reply by chef-SAbhinay** [reply]
> oh! it is mistake.  i have corrected later but still not detection. 

is this resolved?

**Comment by ghost** [comment]
The following code works for me.

```
import torch
model = torch.hub.load('.', 'custom', path='/path/to/yolov5/runs/train/exp5/weights/best.pt', source='local') 
# Image
img = '/path/to/test/image/25.jpg'
# Inference
results = model(img)
# Results, change the flowing to: results.show()
results.show()  # or .show(), .save(), .crop(), .pandas(), etc
```

**Reply by medinikb** [reply]
I could run custom model as below:
`model = torch.hub.load("yolov5/torch/hub/ultralytics_yolov5_master", 'custom', path="best.pt", source='local') `

**Reply by jmiller-dr** [reply]
Suggest to append this info to the Docs. They seem to be overly focused on non-custom inference only.

**Reply by akashlinux10may** [reply]
model = torch.hub.load('.', 'custom', path='/path/to/yolov5/runs/train/exp5/weights/best.pt', source='local') 
This is not working.
Got error
FileNotFoundError: [Errno 2] No such file or directory: '.\\hubconf.py'

**Reply by tudoanh** [reply]
> model = torch.hub.load('.', 'custom', path='/path/to/yolov5/runs/train/exp5/weights/best.pt', source='local') This is not working. Got error FileNotFoundError: [Errno 2] No such file or directory: '.\hubconf.py'

The first param (`"."` in this case) should be point to your local `yolov5` directory

**Reply by akashlinux10may** [reply]
Can I use yolov5 without yolov5 director? because while I am deploying model then hubconf.py and experimental.py files of the folder shows high severity warning.

**Reply by tudoanh** [reply]
> Can I use yolov5 without yolov5 director? because while I am deploying model then hubconf.py and experimental.py files of the folder shows high severity warning.

Of course. You may try to [fetch a pretrained model](https://pytorch.org/docs/stable/hub.html#loading-models-from-hub) from github like this:

```
model = torch.hub.load("ultralytics/yolov5", "yolov5s", pretrained=True)
ckpt = torch.load(model_path)
model.load_state_dict(ckpt['model'].state_dict())
```

**Reply by kessmith** [reply]
> > model = torch.hub.load('.', 'custom', path='/path/to/yolov5/runs/train/exp5/weights/best.pt', source='local') This is not working. Got error FileNotFoundError: [Errno 2] No such file or directory: '.\hubconf.py'
> 
> The first param (`"."` in this case) should be point to your local `yolov5` directory

Do I need to point to a Yolov5 directory to get this code to work?

**Reply by zkytony** [reply]
@tudoanh wait, how to do this for a custom model then? I want to load a pretrained custom model saved in a local `.pt` file.

Previously, `torch.hub.load('ultralytics/yolov5', 'custom', path='path/to/model.pt')` would just work. But now, I am getting all sorts of errors trying to make it work.

**Reply by ammarsaf** [reply]
Your code is so helpful @matrix246, thanks! I just want to know, based on the last line that `results.show()  # or .show(), .save(), .crop(), .pandas(), etc`, is there any option that I can save the `bbox text` from the output?

**Reply by BGAKSHAYA** [reply]
@Ammar-Azman The below code gives the bounding box coordinates, class label and the confidence score for the first image.
`results.pandas().xyxy[0]`

Sample Output:
```
          xmin       ymin        xmax        ymax  confidence  class  name
0   720.201294   4.749833  722.742065  256.000000    0.984483      0  dog
1   739.679504   8.926133  742.159546  256.000000    0.953222      0  dog
```

For xywh format, you can use `results.pandas().xywh[0]`

**Comment by Brunettow** [comment]
I don't want to use custom. However this doesn't work either    
model = torch.hub.load('/home/project/yolov5', 'yolov5x6', path='/home/project/yolov5/best.pt', source='local')

**Reply by pinnintipraneethkumar** [reply]
Hi, @Brunettow you are using it in a wrong way insted of 'yolov5x6' use **'custom'** it will work without any problem. you are trying to laod a custom model so you should use custom.
```
model = torch.hub.load('/home/project/yolov5', 'custom', path='/home/project/yolov5/best.pt', source='local') 
```
